Source/WebKit/UIProcess/mac/DisplayLinkMac.mm

```diff
@@ -0,0 +1,155 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "DisplayLink.h"
+
+#if HAVE(CA_DISPLAY_LINK_MAC)
+
+#import <pal/spi/cocoa/QuartzCoreSPI.h>
+#include <wtf/ProcessPrivilege.h>
+#include <wtf/text/TextStream.h>
+
+@interface WKDisplayLinkHandler : NSObject {
+    WebKit::DisplayLink* _client;
+    RetainPtr<CADisplayLink> _displayLink;
+    RetainPtr<NSThread> _thread;
+}
+
+- (id)initWithScreen:(NSScreen *)screen client:(WebKit::DisplayLink*)client;
+- (void)invalidate;
+- (WebCore::FramesPerSecond)nominalFramesPerSecond;
+
+@property (getter=isRunning) BOOL running;
+
+@end
+
+@implementation WKDisplayLinkHandler
+
+- (id)initWithScreen:(NSScreen *)screen client:(WebKit::DisplayLink*)client
+{
+    if (self = [super init]) {
+        _client = client;
+
+        _displayLink = [screen displayLinkWithTarget:self selector:@selector(displayLinkFired:)];
+        [_displayLink setPaused:YES];
+
+        _thread = adoptNS([[NSThread alloc] initWithTarget:self selector:@selector(threadWasCreated) object:nil]);
+        [_thread setQualityOfService:NSQualityOfServiceUserInteractive];
+        [_thread start];
+    }
+    return self;
+}
+
+- (void)threadWasCreated
+{
+    [_displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+    [[NSRunLoop currentRunLoop] run];
+}
+
+- (void)dealloc
+{
+    ASSERT(!_displayLink);
+    ASSERT(!_thread);
+    [super dealloc];
+}
+
+- (void)displayLinkFired:(CADisplayLink *)sender
+{
+    ASSERT(!isMainRunLoop());
+    _client->displayLinkHandlerCallbackFired();
+}
+
+- (void)invalidate
+{
+    [_displayLink invalidate];
+    _displayLink = nil;
+
+    [_thread cancel];
```

Comment:
This doesn't do anything except set a flag. To actually kick the runloop out of its wait loop, you would need something like `CFRunLoopStop` here (and `CFRunLoopRun()` in the background thread, since `CFRunLoopStop` doesn't do anything for an `-[NSRunLoop run]` invocation due to differing semantics of those calls).

Right now, even after you invalidate the CADisplayLink, the worker thread spawned will just run forever.